File: spec/sessions/support/dice_bridge.mjs
// Scripted-dice bridge for the Session Tests (docs/project/session_tests.md).
//
// The browser is where dice are actually rolled: the Action Builder asks
// Check Resolution for each Roll's TN (after cross-side Propagation), the
// Roll Controller rolls the dice, and Scoring turns them into the
// Successes the page POSTs back. A Session Test scripts the dice instead
// of rolling them, but runs them through *these same modules* so a
// scenario's Successes are exactly what the DM would have seen.
//
// Protocol: one JSON request per line on stdin, one JSON response per
// line on stdout.
//
//   in  { "rolls": [ { "id", "side", "base_tn", "bonus_penalty_list",
//                      "no_propagate", "dice_count", "dice" } ],
//         "spread": false }
//   out { "ok": true, "rolls": { "<id>": { "tn", "starting_value",
//                                          "dois", "critical_count",
//                                          "outcome", "final_dice" } } }
//
// Mirrors public/js/ui/actionBuilder.js#_previewTns (TN) and
// public/js/ui/rollController.js#rollGroup (scoring). Reroll and Nudge are
// deliberately not applied — no scenario spends Luck yet; see the Gaps
// section of docs/project/session_tests.md.

import { createInterface } from 'node:readline';
import { CheckResolution } from '../../../public/js/check.js';
import { Scoring } from '../../../public/js/scoring.js';
import { DiceConfig } from '../../../public/js/config.js';

const config = DiceConfig.default();

function toRoll(spec) {
  return {
    _id: spec.id,
    _dice: spec.dice || [],
    _failureModifier: spec.failure_modifier,
    _criticalModifier: spec.critical_modifier,
    side: spec.side,
    baseTn: spec.base_tn,
    bonusPenaltyList: spec.bonus_penalty_list || [],
    noPropagate: spec.no_propagate || [],
    startingContribution: spec.starting_contribution || 0,
  };
}

function handle(request) {
  const rolls = (request.rolls || []).map(toRoll);
  const supporting = rolls.filter((r) => r.side !== 'opposing');
  const opposing = rolls.filter((r) => r.side === 'opposing');
  const preview = CheckResolution.previewParameters({
    supporting,
    opposing,
    spread: !!request.spread,
  });

  const out = {};
  const collect = (list, results) => {
    list.forEach((roll, i) => {
      const res = results[i];
      if (!res) return;
      const failureModifier =
        roll._failureModifier != null ? roll._failureModifier : config.defaultFailureModifier;
      const criticalModifier =
        roll._criticalModifier != null ? roll._criticalModifier : config.defaultCriticalModifier;
      const score = Scoring.score(
        roll._dice,
        {
          tn: res.tn,
          startingValue: res.startingValue,
          failureModifier,
          criticalModifier,
        },
        config,
      );
      out[roll._id] = {
        tn: res.tn,
        starting_value: res.startingValue,
        dois: score.dois,
        critical_count: score.criticalCount,
        outcome: score.outcome,
        final_dice: roll._dice,
      };
    });
  };
  collect(supporting, preview.supporting);
  collect(opposing, preview.opposing);
  return { ok: true, rolls: out };
}

const rl = createInterface({ input: process.stdin });
rl.on('line', (line) => {
  if (!line.trim()) return;
  let response;
  try {
    response = handle(JSON.parse(line));
  } catch (err) {
    response = { ok: false, error: String((err && err.message) || err) };
  }
  process.stdout.write(JSON.stringify(response) + '\n');
});

File: spec/sessions/support/dice_bridge.rb
require 'json'
require 'open3'

module Sessions
  # Ruby side of the scripted-dice bridge. Keeps one long-lived `node`
  # process running dice_bridge.mjs so a Session Test resolves its Rolls
  # through the browser's own Check Resolution / Scoring modules instead
  # of a Ruby re-implementation of them.
  #
  # See docs/project/session_tests.md — "Scripted dice".
  module DiceBridge
    SCRIPT = File.expand_path('dice_bridge.mjs', __dir__).freeze

    class Unavailable < StandardError; end
    class Error < StandardError; end

    module_function

    # True when a `node` capable of running the bridge is on PATH. Session
    # specs skip themselves (rather than fail) when it is missing.
    def available?
      return @available unless @available.nil?
      @available = system('node', '--version', out: File::NULL, err: File::NULL) ? true : false
    end

    # Resolve a set of Rolls against their scripted dice.
    #
    #   rolls  — [{ id:, side:, base_tn:, bonus_penalty_list:, dice_count:,
    #               dice: [...] }, ...]
    #   spread — true for an area (Spread) Check: the caster is compared
    #            against each Opposer independently.
    #
    # Returns { "<id>" => { tn:, starting_value:, dois:, critical_count:,
    #                       outcome:, final_dice: } }.
    def resolve(rolls:, spread: false)
      request = { rolls: rolls.map { |r| stringify(r) }, spread: spread }
      response = exchange(request)
      raise Error, "dice bridge: #{response['error']}" unless response['ok']
      response['rolls'].transform_values { |v| symbolize(v) }
    end

    def exchange(request)
      io = process
      io[:stdin].puts(JSON.generate(request))
      io[:stdin].flush
      line = io[:stdout].gets
      raise Error, 'dice bridge closed unexpectedly' if line.nil?
      JSON.parse(line)
    end

    def process
      @process ||= begin
        raise Unavailable, 'node is required to run the Session Tests' unless available?
        stdin, stdout, stderr, wait = Open3.popen3('node', SCRIPT)
        at_exit { shutdown }
        { stdin: stdin, stdout: stdout, stderr: stderr, wait: wait }
      end
    end

    def shutdown
      return unless defined?(@process) && @process
      @process[:stdin].close rescue nil
      @process[:wait].join   rescue nil
      @process = nil
    end

    def stringify(roll)
      roll.each_with_object({}) { |(k, v), h| h[k.to_s] = v }
    end

    def symbolize(hash)
      hash.each_with_object({}) { |(k, v), h| h[k.to_sym] = v }
    end
  end
end
